src/wp-includes/html-api/class-wp-html-active-formatting-elements.php

```diff
@@ -0,0 +1,183 @@
+	/**
+	 * Holds the stack of active formatting element references.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @var WP_HTML_Token[]
+	 */
+	private $stack = array();
+
+	/**
+	 * Reports if a specific node is in the stack of active formatting elements.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param WP_HTML_Token $token Look for this node in the stack.
+	 * @return bool Whether the referenced node is in the stack of active formatting elements.
+	 *
+	 */
+	public function contains_node( $token ) {
+		foreach ( $this->walk_up() as $item ) {
+			if ( $token->bookmark_name === $item->bookmark_name ) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	/**
+	 * Returns how many nodes are currently in the stack of active formatting elements.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @return int How many node are in the stack of active formatting elements.
+	 */
+	public function count() {
+		return count( $this->stack );
+	}
+
+	/**
+	 * Returns the node at the end of the stack of active formatting elements,
+	 * if one exists. If the stack is empty, returns null.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @return WP_HTML_Token|null Last node in the stack of active formatting elements, if one exists, otherwise null.
+	 */
+	public function current_node() {
+		$current_node = end( $this->stack );
+
+		return $current_node ? $current_node : null;
+	}
+
+	/**
+	 * Pushes a node onto the stack of active formatting elements.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @see https://html.spec.whatwg.org/#push-onto-the-list-of-active-formatting-elements
+	 *
+	 * @param WP_HTML_Token $token Push this node onto the stack.
+	 */
+	public function push( $token ) {
+		/*
+		 * > If there are already three elements in the list of active formatting elements after the last marker,
+		 * > if any, or anywhere in the list if there are no markers, that have the same tag name, namespace, and
+		 * > attributes as element, then remove the earliest such element from the list of active formatting
+		 * > elements. For these purposes, the attributes must be compared as they were when the elements were
+		 * > created by the parser; two elements have the same attributes if all their parsed attributes can be
+		 * > paired such that the two attributes in each pair have identical names, namespaces, and values
+		 * > (the order of the attributes does not matter).
+		 *
+		 * @TODO: Implement the "Noah's Ark clause" to only add up to three of any given kind of formatting elements to the stack.
+		 */
+
+		// > Add element to the list of active formatting elements.
+		$this->stack[] = $token;
+	}
+
+	/**
+	 * Removes a node from the stack of active formatting elements.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param WP_HTML_Token $token Remove this node from the stack, if it's there already.
+	 */
+	public function remove_node( $token ) {
+		foreach ( $this->walk_up() as $position => $item ) {
+			if ( $token->bookmark_name === $item->bookmark_name ) {
+				array_splice( $this->stack, $this->count() - $position - 1, 1 );
```

Comment:
(Recurring) nit -- could do negative offset instead of subtraction:

```suggestion
				array_splice( $this->stack, - $position - 1, 1 );
```

Not sure it's that much nicer 🤷‍♂️

Reply:
hm. maybe if others chime in. I'm confused more by negative positions, particularly when they also include math. I with the function would reject negative offsets so that subtraction and negative offsets couldn't overlap.

Reply:
I think a negative offset is harder to read. What about `1 - $position`?

Reply:
I think a negative offset is harder to read. What about `1 - $position`? It would avoid the call to `$this->count()`, and while easy to read, I'm not sure how if it would be as clear to others.

Reply:
I think a negative offset with `- $position - 1` is harder to read. What about `1 - $position`? It would avoid the call to `$this->count()`, and while it's personally easier for me to read, I'm not sure if it would be as clear to others.

Reply:
`1 - $position` is even less clear to me. I'm not worried about `$this->count()` as a call. although in this case it's a function call, we could decide if we measure that it's a problem to switch it to `count( $this->stack )` which is essentially the same as a normal variable lookup.

Reply:
I've rearranged this a bit to try and help. the loop index is now called `$position_from_end` to better indicate that it's not the normal array index. we're also using a helper variable `$position_from_start` to give a better name to the computation.